network/phonebook.go

```diff
@@ -114,24 +114,36 @@ func (p *ThreadsafePhonebook) ReplacePeerList(they []string) {
 	copy(p.addrs, they)
 }
 
-// MultiPhonebook contains several phonebooks
+// MultiPhonebook contains a map of phonebooks
 type MultiPhonebook struct {
-	phonebooks []Phonebook
-	lock       deadlock.RWMutex
+	phonebookMap map[string]*Phonebook
+	lock         deadlock.RWMutex
+}
+
+// MakeMultiPhonebook constructs and returns a new Multi Phonebook
+func MakeMultiPhonebook() *MultiPhonebook {
+	return &MultiPhonebook{phonebookMap: make(map[string]*Phonebook)}
 }
 
 // GetAddresses returns up to N address
 // TODO: this implementation does a bunch of extra copying, make it more efficient
 func (mp *MultiPhonebook) GetAddresses(n int) []string {
 	mp.lock.RLock()
 	defer mp.lock.RUnlock()
-	if len(mp.phonebooks) == 1 {
-		return mp.phonebooks[0].GetAddresses(n)
+
+	phonebookList := []Phonebook{}
+
+	for _, phonebook := range mp.phonebookMap {
+		phonebookList = append(phonebookList, *phonebook)
 	}
-	sizes := make([]int, len(mp.phonebooks))
+
+	if len(phonebookList) == 1 {
+		return phonebookList[0].GetAddresses(n)
+	}
+	sizes := make([]int, len(phonebookList))
 	total := 0
-	addrs := make([][]string, len(mp.phonebooks))
-	for pi, p := range mp.phonebooks {
+	addrs := make([][]string, len(phonebookList))
+	for pi, p := range phonebookList {
```

Comment:
We don't need a phonebookList array at all - just iterate over phonebookMap, and put the `len(map) == 0` check in this one loop.

Reply:
removed phonebookList array.